pkg/webhook/mutation.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation"
+	"github.com/open-policy-agent/gatekeeper/pkg/util"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	clientcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type mutationResponse string
+
+const (
+	mutationSkipResponse    mutationResponse = "skip"
+	mutationUnknownResponse mutationResponse = "unknown"
+)
+
+func init() {
+	AddToManagerFuncs = append(AddToManagerFuncs, AddMutatingWebhook)
+
+	if err := apis.AddToScheme(runtimeScheme); err != nil {
+		log.Error(err, "unable to add to scheme")
+		panic(err)
+	}
+}
+
+// +kubebuilder:webhook:verbs=create,path=/v1/mutate,mutating=true,failurePolicy=ignore,groups=*,resources=*,versions=*,name=mutation.gatekeeper.sh
+// +kubebuilder:rbac:groups=*,resources=*,verbs=get;list;watch;update
+
+// AddMutatingWebhook registers the mutating webhook server with the manager
+func AddMutatingWebhook(mgr manager.Manager, client *opa.Client, processExcluder *process.Excluder) error {
+	if !*mutation.MutationEnabled {
+		return nil
+	}
+	reporter, err := newStatsReporter()
+	if err != nil {
+		return err
+	}
+	eventBroadcaster := record.NewBroadcaster()
+	kubeClient := kubernetes.NewForConfigOrDie(mgr.GetConfig())
+
+	eventBroadcaster.StartRecordingToSink(&clientcorev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(
+		scheme.Scheme,
+		corev1.EventSource{Component: "gatekeeper-mutation-webhook"})
+
+	wh := &admission.Webhook{
+		Handler: &mutationHandler{
+			webhookHandler: webhookHandler{
+				client:          mgr.GetClient(),
+				reader:          mgr.GetAPIReader(),
+				reporter:        reporter,
+				processExcluder: processExcluder,
+				eventRecorder:   recorder,
+				gkNamespace:     util.GetNamespace(),
+			},
+		},
+	}
+
+	// TODO(https://github.com/open-policy-agent/gatekeeper/issues/661): remove log injection if the race condition in the cited bug is eliminated.
+	// Otherwise we risk having unstable logger names for the webhook.
+	if err := wh.InjectLogger(log); err != nil {
+		return err
+	}
+	mgr.GetWebhookServer().Register("/v1/mutate", wh)
+
+	return nil
+}
+
+var _ admission.Handler = &mutationHandler{}
+
+type mutationHandler struct {
+	webhookHandler
+}
+
+// Handle the mutation request
+func (h *mutationHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log := log.WithValues("hookType", "mutation")
+	var timeStart = time.Now()
+
+	if isGkServiceAccount(req.AdmissionRequest.UserInfo) {
+		return admission.ValidationResponse(true, "Gatekeeper does not self-manage")
+	}
+
+	if req.AdmissionRequest.Operation != admissionv1beta1.Create &&
+		req.AdmissionRequest.Operation != admissionv1beta1.Update {
```

Comment:
need to support update here

Reply:
Updated reponse comment (is this what you mean?)

Reply:
ah yes I see the update now